### PR TITLE
Adding user_data to avl interface so that an exec_ctx can be passed in.

### PR DESCRIPTION
**Pull request by yang-g**
This does not completely fix the issue in https://github.com/grpc/grpc/pull/11433. But might be an improvement to the current subchannel_index implementation.

### CODE REVIEW

File: src/core/lib/support/avl.c [hunk 1]
```diff
@@ -32,22 +32,23 @@ gpr_avl gpr_avl_create(const gpr_avl_vtable *vtable) {
   return out;
 }
 
-static gpr_avl_node *ref_node(gpr_avl_node *node) {
+static gpr_avl_node *ref_node(gpr_avl_node *node, void *user_data) {
```
Comment (markdroth) [inline src/core/lib/support/avl.c]: Why add the `user_data` param here if it's not used for anything?

Comment (yang-g) [inline src/core/lib/support/avl.c]: Done.

**Comment by yang-g** [comment]
This PR solves the problem of maintaining the life time of a thread local exec ctx. It can be cleared in the shutdown call yet the shutdown call only unref the avl. If some of the avl vtable function is called afterwards (e.g. destroy), it will try to grab the destroyed tls exec_ctx.

File: include/grpc/support/avl.h [hunk 2]
```diff
@@ -57,25 +62,27 @@ typedef struct gpr_avl {
 GPRAPI gpr_avl gpr_avl_create(const gpr_avl_vtable *vtable);
 /** add a reference to an existing tree - returns
```
Comment (dgquintas) [inline include/grpc/support/avl.h]: We should update the docstring for all the modified functions that now take `user_data`

Comment (yang-g) [inline include/grpc/support/avl.h]: Done.

**Comment by yang-g** [comment]
PTAL @dgquintas Thanks.

**Pull request merged**